### PR TITLE
Deleting Unhealthy game servers

### LINKED ISSUE #56: Action when the GameServer is Unhealthy
A GameServer instance can turn Unhealthy on two occasions.
- The game server process can set itself in Unhealthy from inside GSDK
- The NodeAgent can detect an absence in heartbeats.

Currently we don't do anything when this happens. We should think if we can do something (like log on the GameServer controller and delete the GameServer instance?).

### LINKED ISSUE #256: Take into account Unhealthy GameServers when scaling
In scenarios that we have Unhealthy GameServers, GameServerBuild controller does not take them into account when scaling.

### PR DESCRIPTION
**Pull request by dgkanatsios**
Fixes #56 

This PR modifies the GameServerBuild controller to delete a GameServer when its status becomes Unhealthy.

As a side effect, we expect it to fix #256 and #257 since Unhealthy GameServers won't exist in the cluster any more so:

i) they won't be taken into account when scaling
ii) NodeAgent will get a notification about their deletion so it will not repeatedly try to mark them as Unhealthy

The change in the YAML files is probably due to the new C-R version that was merged in a different PR.

### CODE REVIEW

File: pkg/operator/controllers/metrics.go [hunk 1]
```diff
@@ -40,6 +40,14 @@ var (
 		},
 		[]string{"BuildName"},
 	)
+	GameServersUnhealthyCounter = registry.NewCounterVec(
```
Comment (javier-op) [inline pkg/operator/controllers/metrics.go:43]: Could we add this to the table at the bottom of the monitoring section?
https://github.com/PlayFab/thundernetes/blob/main/docs/howtos/monitoring.md

File: pkg/operator/controllers/gameserverbuild_controller.go [hunk 2]
```diff
@@ -142,27 +142,35 @@ func (r *GameServerBuildReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
-		} else if gs.Status.State == mpsv1alpha1.GameServerStateInitializing {
+		} else if gs.Status.State == mpsv1alpha1.GameServerStateInitializing && gs.Status.Health == mpsv1alpha1.GameServerHealthy {
 			initializingCount++
-		} else if gs.Status.State == mpsv1alpha1.GameServerStateStandingBy {
+		} else if gs.Status.State == mpsv1alpha1.GameServerStateStandingBy && gs.Status.Health == mpsv1alpha1.GameServerHealthy {
 			standingByCount++
-		} else if gs.Status.State == mpsv1alpha1.GameServerStateActive {
+		} else if gs.Status.State == mpsv1alpha1.GameServerStateActive && gs.Status.Health == mpsv1alpha1.GameServerHealthy {
 			activeCount++
+		} else if gs.Status.State == mpsv1alpha1.GameServerStateGameCompleted && gs.Status.Health == mpsv1alpha1.GameServerHealthy {
```
Comment (javier-op) [inline pkg/operator/controllers/gameserverbuild_controller.go:151]: Just a question, this case is for when the GameServer finished the session and exited gracefully?

File: cmd/e2e/build_unhealthy_gameservers_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"context"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	mpsv1alpha1 "github.com/playfab/thundernetes/pkg/operator/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Build with unhealthy game servers", func() {
```
Comment (javier-op) [inline cmd/e2e/build_unhealthy_gameservers_test.go:13]: Could we change the description for "should delete unhealthy gameservers and create new ones" or something like that?

File: pkg/operator/controllers/gameserverbuild_controller.go (hunk 2 above)
Comment (dgkanatsios) [reply pkg/operator/controllers/gameserverbuild_controller.go:151]: yup and I think that this is the case when Kubernetes evicts Pods, so they get an exit code of zero so they transition to the Completed state

File: cmd/e2e/build_unhealthy_gameservers_test.go (hunk 3 above)
Comment (dgkanatsios) [reply cmd/e2e/build_unhealthy_gameservers_test.go:13]: will do, thanks

**Review by javier-op — APPROVED** [review]

File: pkg/operator/controllers/metrics.go (hunk 1 above)
Comment (dgkanatsios) [reply pkg/operator/controllers/metrics.go:43]: tracked in #280

**Pull request merged**